### ISSUE TITLE
Block flatdb has too many buckets.

### ISSUE THREAD

**Issue by mateon1** [ipfs/go-ipfs#3062]
Version/Platform/Processor information: 0.4.2/Windows 10 x64

Type: bug/enchancement
Area: blockstore
Priority: P0/P1

Description:
The current flatdb format for blocks has way too many buckets. Most filesystems slow down drastically as the number of objects in a directory increases.

Git uses 2 hexadecimal characters for buckets. I believe 2 or 3 is the most a filesystem like Windows's NTFS can handle without drastic slowdown.
Alternatively, the blockdb could use several directories, as in blocks/1a/2b/3c4d5e...9f.data

When adding a lot of objects, for example a 20 gigabyte file, in the current system of 6 characters per bucket, there is an average of 2 .data files per bucket, and the amount of buckets themselves slow down block operations by at least a factor of 8 (NTFS)

**Comment by mateon1** [comment]
Related: Repo GC doesn't clean up empty directories in the block database.

**Comment by djdv** [comment]
Adding an anecdote here, I've run into a "Too many links" issue on a FreeBSD system before when adding a lot of files to ipfs on UFS. For reference it was an old 32bit testing machine running I think fbsd 9 and ipfs 3.x at the time. It couldn't handle all the directories under blocks/.

**Comment by Kubuxu** [comment]
Which version are you running?

There is difference in total prefix length and amount of entropy in that prefix, due to multihash prefix. In go-ipfs 0.4.3-X there are 512 buckets in the flatfs. Which is only bit more than git uses. This is also different from what was used in 0.4.2 and before.

**Comment by mateon1** [comment]
@Kubuxu I'm using go-ipfs version 0.4.2 (`version --all` fails on this one), the prefix length in my database is 8 hex characters, with the first 4 of them being constant.
I'm not sure how to install and test the 0.4.3 version on Windows, but I'll look in the docs and ask on IRC.

**Comment by jbenet** [comment]
Yeah I warned about this before. I think the verdict was "any filesystem
will do fine on this, you're worrying ove nothing".

We can implement a migration to make the fanout smaller and deeper. Though
if there's serious perf implications wonder if it should be a config
dependent thing. That May be error prone. Maybe just change fanout and gun
for arena storage.

**Comment by Kubuxu** [comment]
@jbenet it might not be an issue after all, we already migrated from 16bit fanout to 5bit limiting number of buckets from 65k to 512.

**Comment by mateon1** [comment]
@jbenet Actually, in 0.4.2, this was a huge problem, with all the buckets causing massive slowdowns on the order of 8-10x slower than an empty repo. It's way less of a problem in 0.4.3, but it still doesn't scale to terabytes in the repo.

**Comment by Kubuxu** [comment]
Yes for that we would need to implement variable fanout depending on number of blocks stored in previous level. It will be complex as you need to both take into account expanding and contracting the fanout.

**Comment by jbenet** [comment]
> Yes for that we would need to implement variable fanout depending on number of blocks stored in previous level. It will be complex as you need to both take into account expanding and contracting the fanout.

We could just expand for now. implement compaction when it is needed

**Comment by jbenet** [comment]
but yes, we really need it to be variable sized. [guess who implemented this years ago](https://github.com/datastore/datastore/blob/7ccf0cd4748001d3dbf5e6dda369b0f63e0269d3/datastore/core/basic.py#L591-L673).
